### PR TITLE
CSPL-624-Splunk-Operator-upgrade-documentation

### PR DESCRIPTION
**Pull request by pdhanoyasplunk**

### CODE REVIEW

File: docs/SplunkOperatorUpgrade.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
+# How to upgrade Splunk Operator and Splunk Enterprise Deployments
+
+To upgrade Splunk Operator Kubernetes(SKO), you must install new version directly on top of old version(On same Kubernetes Cluster/Namespace). When Splunk Operator Uprades it updates CRDs and Splunk PODs deployed by the Operator to match the latest SPEC. An upgrade of the Splunk Operator will initiate a restart of all deployed pods. If the Splunk Enterprise Docker image changes, the deployed pods will be upgraded and restarted. 
+In a typical Splunk Multisite Indexer Cluster with Search Head Cluster the upgrade will follow the below order:
+* A new Splunk Operator pod will be created and the existing operator pod will drop
+* If a License Master is present. License Master pod will be terminated and new pod with updated Spec will be created.
+* After new License Master pod comes up Cluster Master and Deployer pods will be terminated
+* New Cluster Master and Deployer pod are crated with updated spec
```
Comment (smohan-splunk) [inline docs/SplunkOperatorUpgrade.md:8]: created

File: docs/SplunkOperatorUpgrade.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+* After Cluster Master Pod is ready, Indexer Pods are reset one by one per site. 
+* After Deployer Pod is ready, Search Heads are reset one by one.
```
Comment (smohan-splunk) [inline docs/SplunkOperatorUpgrade.md:10]: Can SHs be reset while the Indexers are being rest?

File: docs/SplunkOperatorUpgrade.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+* Once all the members of Indexer Cluster and Search Head Cluster are ready, Monitoring Console Pod will be reset.
+
+## Prequisites 
+
+Before you upgrade, review the Splunk Operator [change log](https://github.com/splunk/splunk-operator/releases) page for information on changes made in the latest release.
+
+The Splunk Operator does not provide a means for downgrading to a previous version. If you need to revert to a prior Splunk Operator release, remove the current release and reinstall the version you want.
```
Comment (smohan-splunk) [inline docs/SplunkOperatorUpgrade.md:17]: "If you need to revert to a prior Splunk Operator release, remove the current release and reinstall the version you want."
We should explain the implications of doing this

File: docs/SplunkOperatorUpgrade.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+NAME                               READY   STATUS    RESTARTS   AGE
+splunk-operator-75f5d4d85b-8pshn   1/1     Running   0          5s
+
+```
+
+If a Splunk Operator release changes the custom resource (CRD) API version, the administrator is required to update their CRD yaml files to use the latest CRD API version and apply the changes.
+
+If a Splunk Operator release includes an updated Splunk Enterprise Docker image, the operator will also initiate Splunk Enterprise version update.
+
+## Verifiy Upgrade is Successfull
```
Comment (smohan-splunk) [inline docs/SplunkOperatorUpgrade.md:56]: Verify Upgrade is Successful

File: docs/SplunkOperatorUpgrade.md (hunk 2 above)
Comment (pdhanoyasplunk) [reply docs/SplunkOperatorUpgrade.md:10]: Yes they can be reset side by side.

File: docs/SplunkOperatorUpgrade.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+NAME                               READY   STATUS    RESTARTS   AGE
+splunk-operator-75f5d4d85b-8pshn   1/1     Running   0          5s
+​
+```
+​
+If a Splunk Operator release changes the custom resource (CRD) API version, the administrator is responsible for updating their Custom Resource specification to reference the latest CRD API version.
+​
+If a Splunk Operator release includes an updated Splunk Enterprise Docker image, the operator upgrade will also initiate pod restart using the latest Splunk Enterprise Docker image.
+​
+## Verifiy Upgrade is Successful
```
Comment (smohan-splunk) [inline docs/SplunkOperatorUpgrade.md:48]: "Verify"

File: docs/SplunkOperatorUpgrade.md [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+kubectl get pods splunk-default-monitoring-console-0 -o yaml | grep -i image
+image: splunk/splunk:8.1.2
+imagePullPolicy: IfNotPresent
+image: splunk/splunk:8.1.2
+```
+​
+## Splunk Enterprise Cluster upgrade example
+This is an example of the process followed by the Splunk Operator if the operator version is upgraded and a later Splunk Enterprise Docker image is available:
+​
+1. A new Splunk Operator pod will be created, and any existing operator pod will drop.
```
Comment (smohan-splunk) [inline docs/SplunkOperatorUpgrade.md:70]: "wil be dropped"

**Review by smohan-splunk — APPROVED** [review]

File: docs/SplunkOperatorUpgrade.md [hunk 7]
```diff
@@ -0,0 +1,77 @@
+# How to upgrade Splunk Operator and Splunk Enterprise Deployments
+​
+To upgrade the Splunk Operator for Kubernetes, you will overwrite the prior operator release with the latest version. Once the latest version of the Splunk Operator is running, the CRD's are updated and any running pods in the operator managed namespace are restarted. Any new spec defined by the operator will be applied to the restarted pods.
```
Comment (akondur) [inline docs/SplunkOperatorUpgrade.md:3]: IMO this is the order in which we should do the update:

1. Update the CRD's
2. Update the operator deployment with a newer operator docker image version

Note: The **splunk-operator-install.yaml** also applies the CRDs before the deployment for the operator.

After step 1 above the K8S cluster, ETCD supports and stores both versions(old and new) versions of the CRD(i.e any CRs already deployed with either version) with the way we have structured our CRD's. But any new changes bought in with the new CRD will not be acted upon by the older version of the operator as the operator doesn't have the capability to understand and act upon the same(which is fine as older CRs with an older apiVersion for the CRD will still be served)

After step 2, the operator is well equipped to handle any such changes to the CRD.

If we reverse the order of the steps above, the Splunk Operator might be looking for some spec within a CR that has been deployed already and might not exist(the effects of which might not be catastrophic but rather unknown).

In general, having both versions of the CRDs stored in the database before the operator is upgraded might serve well for both existing deployments, future upgrades/deployments.

Would love to hear other thoughts if any.

File: docs/SplunkOperatorUpgrade.md [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+* For general information about Splunk Enterprise compatibility and the upgrade process, see [How to upgrade Splunk Enterprise](https://docs.splunk.com/Documentation/Splunk/latest/Installation/HowtoupgradeSplunk).
+​
+* If you use forwarders, verify the Splunk Enterprise version compatibility with the forwarders in the [Compatibility between forwarders and Splunk Enterprise indexers](https://docs.splunk.com/Documentation/Forwarder/latest/Forwarder/Compatibilitybetweenforwardersandindexers) documentation.
+​
+## Upgrading Splunk Operator and Splunk Operator Deployment
+​
+1. Download the latest Splunk Operator installation yaml file.
+​
+```
+wget -O splunk-operator.yaml https://github.com/splunk/splunk-operator/releases/download/0.2.2/splunk-operator-install.yaml
```
Comment (akondur) [inline docs/SplunkOperatorUpgrade.md:24]: Nit: **`wget -O splunk-operator.yaml https://github.com/splunk/splunk-operator/releases/download/<desired_version_of_operator>/splunk-operator-install.yaml `**?

File: docs/SplunkOperatorUpgrade.md [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+​
+2. (Optional) Review the file and update it with your specific customizations used during your install. 
+​
+3. Upgrade the Splunk Operator.
+​
+```
+kubectl apply -f splunk-operator.yaml
+```
+​
+After applying the yaml, a new operator pod will be created and the existing operator pod will drop. Example:
```
Comment (akondur) [inline docs/SplunkOperatorUpgrade.md:35]: Nit: **`will be terminated`** vs **`will drop`**?

File: docs/SplunkOperatorUpgrade.md [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+​
+```
+​
+If a Splunk Operator release changes the custom resource (CRD) API version, the administrator is responsible for updating their Custom Resource specification to reference the latest CRD API version.
+​
+If a Splunk Operator release includes an updated Splunk Enterprise Docker image, the operator upgrade will also initiate pod restart using the latest Splunk Enterprise Docker image.
+​
+## Verifiy Upgrade is Successful
+​
+To verify the Splunk Operator has been upgraded to the release image in splunk-operator-install.yaml, you can check the version of the operator image. Example:
```
Comment (akondur) [inline docs/SplunkOperatorUpgrade.md:50]: To verify the Splunk Operator has been upgraded to the desired version indicated in **`splunk-operator-install.yaml`**, you can check the version of the operator image in the deployment spec and subsequently the image in Pod spec of the newly deployed operator pod.

File: docs/SplunkOperatorUpgrade.md [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+​
+```
+kubectl get pods splunk-operator-6bcb8dd8f4-glkxt -o yaml | grep -i image
```
Comment (akondur) [inline docs/SplunkOperatorUpgrade.md:53]:
```
kubectl get deployment splunk-operator -o yaml | grep -i image
image: docker.io/splunk/splunk-operator:<desired_operator_version>
imagePullPolicy: IfNotPresent
``` 

```
kubectl get pod <splunk_operator_pod> -o yaml | grep -i image
image: docker.io/splunk/splunk-operator:<desired_operator_version>
imagePullPolicy: IfNotPresent 
```

The deployment name is always splunk-operator vs the pod name which has a newly generated hash in its name which might be hard to depict in documentation.

File: docs/SplunkOperatorUpgrade.md (hunk 6 above)
Comment (akondur) [reply docs/SplunkOperatorUpgrade.md:70]: will be terminated?

File: docs/SplunkOperatorUpgrade.md [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+image: splunk/splunk:8.1.2
+imagePullPolicy: IfNotPresent
+image: splunk/splunk:8.1.2
+```
+​
+## Splunk Enterprise Cluster upgrade example
+This is an example of the process followed by the Splunk Operator if the operator version is upgraded and a later Splunk Enterprise Docker image is available:
+​
+1. A new Splunk Operator pod will be created, and any existing operator pod will drop.
+2. If a License Master (LM) is present, the LM pod will be dropped.
```
Comment (akondur) [inline docs/SplunkOperatorUpgrade.md:71]: terminated vs dropped?

File: docs/SplunkOperatorUpgrade.md [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+image: splunk/splunk:8.1.2
+```
+​
+## Splunk Enterprise Cluster upgrade example
+This is an example of the process followed by the Splunk Operator if the operator version is upgraded and a later Splunk Enterprise Docker image is available:
+​
+1. A new Splunk Operator pod will be created, and any existing operator pod will drop.
+2. If a License Master (LM) is present, the LM pod will be dropped. 
+3. A new LM pod is started with the updated specifications.
+4. After the new LM pod is up, any Cluster Manager (CM) and Deployer pods will be dropped.
```
Comment (akondur) [inline docs/SplunkOperatorUpgrade.md:73]: I think the order needs to be re-worked here:

1. A new Splunk Operator pod will be created, and the existing operator pod will be terminated.
2. Any existing License Master, Search Head, Deployer, ClusterMaster, Standalone pods will be terminated to be redeployed with the upgraded spec.
3. After a ClusterMaster pod is restarted, the Indexer Cluster pods which are connected to it are terminated and redeployed.
4. After all pods in the Indexer cluster and Search head cluster are redeployed, the Monitoring Console pod is terminated and redeployed.

(Note: If there are multiple pods per Custom Resource, the pods are terminated and re-deployed in a descending order with the highest numbered pod going first)

In a SHC scenario, the Search Head Cluster pods are not dependent on the deployer restart to complete. They act as separate entities. Only the indexer pods are dependent on the ClusterMaster they are connected to i.e they terminate and redeploy only once the ClusterMaster is up and running. @gaurav-splunk @sgontla @kashok-splunk, could you review and see if this is correct?

File: docs/SplunkOperatorUpgrade.md (hunk 7 above)
Comment (pdhanoyasplunk) [reply docs/SplunkOperatorUpgrade.md:3]: Do we want to make this distinction to the customers ? Considering we are offering customer to use splunk-operator-install.yaml which handles upgrade to both CRD and operator?

File: docs/SplunkOperatorUpgrade.md [hunk 14]
```diff
@@ -0,0 +1,77 @@
+# How to upgrade Splunk Operator and Splunk Enterprise Deployments
+​
+To upgrade the Splunk Operator for Kubernetes, you will overwrite the prior operator release with the latest version. Once the latest version of the Splunk Operator is running, the CRD's are updated and any running pods in the operator managed namespace are restarted. Any new spec defined by the operator will be applied to the restarted pods.
+​
+A Splunk Operator for Kubernetes upgrade might include support for a later version of the Splunk Enterprise Docker image. After the Splunk Operator completes its upgrade, the pods will be restarted using the latest Splunk Enterprise Docker image.
```
[the remark this answers is not in the document]
Comment (pdhanoyasplunk) [reply docs/SplunkOperatorUpgrade.md:5]: @eddikay Can you please review this wording?

File: docs/SplunkOperatorUpgrade.md (hunk 8 above)
Comment (pdhanoyasplunk) [reply docs/SplunkOperatorUpgrade.md:24]: That will be a non-working link. I am under the impression we will update this URL with new release. @smohan-splunk Can you please confirm how we want to approach this ?

File: docs/SplunkOperatorUpgrade.md (hunk 9 above)
Comment (pdhanoyasplunk) [reply docs/SplunkOperatorUpgrade.md:35]: That wording was suggested by @eddikay. The consensus is to use terminated @eddikay  can you please confirm its acceptable?

File: docs/SplunkOperatorUpgrade.md (hunk 5 above)
Comment (pdhanoyasplunk) [reply docs/SplunkOperatorUpgrade.md:48]: Will correct.

File: docs/SplunkOperatorUpgrade.md (hunk 11 above)
Comment (pdhanoyasplunk) [reply docs/SplunkOperatorUpgrade.md:53]: This was an example but I can definitely modify to deployment form for consistency.

File: docs/SplunkOperatorUpgrade.md (hunk 13 above)
Comment (pdhanoyasplunk) [reply docs/SplunkOperatorUpgrade.md:73]: You are correct about behavior with Deployer and SHC. Will update the example.

File: docs/SplunkOperatorUpgrade.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (eddikay) [reply docs/SplunkOperatorUpgrade.md:5]: Sure. You can use "In that scenario..."

File: docs/SplunkOperatorUpgrade.md (hunk 7 above)
Comment (eddikay) [reply docs/SplunkOperatorUpgrade.md:3]: I don't understand the comments here. It sounds like the upgrade order of ops is not understood.

Comment (akondur) [reply docs/SplunkOperatorUpgrade.md:3]: We can avoid the distinction completely or have the correct order.

File: docs/SplunkOperatorUpgrade.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (pdhanoyasplunk) [reply docs/SplunkOperatorUpgrade.md:5]: @akondur  *the pods of all Enterprise CR's* wont include Monitoring Console as that is not a CR. I think *all pods managed by Splunk Operator for Kubernete* makes more sense ?

File: docs/SplunkOperatorUpgrade.md [hunk 15]
```diff
@@ -0,0 +1,82 @@
+# How to upgrade Splunk Operator and Splunk Enterprise Deployments
+​
+To upgrade the Splunk Operator for Kubernetes, you will overwrite the prior operator release with the latest version. Once the lastest version of `splunk-operator-install.yaml` is applied the CRD's are updated and operator deployment is updated with newer version of Splunk Operator image. Any new spec defined by the operator will be applied to the pods managed by Splunk Operator for Kubernetes.
```
Comment (pdhanoyasplunk) [inline docs/SplunkOperatorUpgrade.md:3]: Add (see below) after `splunk-operator-install.yaml`

File: docs/SplunkOperatorUpgrade.md (hunk 8 above)
Comment (pdhanoyasplunk) [reply docs/SplunkOperatorUpgrade.md:24]: Leaving at 0.2.2.

File: docs/SplunkOperatorUpgrade.md (hunk 9 above)
Comment (pdhanoyasplunk) [reply docs/SplunkOperatorUpgrade.md:35]: Will be changing to **terminated** as that is the official term by kubernetes.

**Review by akondur — APPROVED** [review]

**Review by eddikay — APPROVED** [review]

**Pull request merged**